thylint/thylint.py

```diff
@@ -0,0 +1,317 @@
+
+
+def regexp_of(words):
+    """Turn word list into regexp.
+
+    Turn a list of word(ish) regexp strings into a compiled regexp that is the big union of
+    whole-word matches, with Isabelle-ish conventions for word boundaries. In particular,
+    count \' as a letter.
+    """
+    return re.compile('|'.join(r'\b(?<!\')(?:{0})\b(?!\')'.format(x) for x in words))
+
+
+diag_words = (
+    # title
+    "Interactive diagnostic command",
+    # message
+    "This command is usually used interactively only and should only be checked in for "
+    "demonstration purposes.",
+    # word(ish) regexps that should be flagged as warnings
+    regexp_of([
+        'ML_val',
+        'class_deps',
+        'code_deps',
+        'code_thms',
+        'find_consts',
+        'find_theorems',
+        'find_unused_assms',
+        'full_prf',
+        'help',
+        'locale_deps',
+        'prf',
+        'print_ML_antiquotations',
+        'print_abbrevs',
+        'print_antiquotations',
+        'print_attributes',
+        'print_bnfs',
+        'print_bundles',
+        'print_case_translations',
+        'print_cases',
+        'print_claset',
+        'print_classes',
+        'print_codeproc',
+        'print_codesetup',
+        'print_coercions',
+        'print_commands',
+        'print_context',
+        'print_definitions',
+        'print_defn_rules',
+        'print_facts',
+        'print_induct_rules',
+        'print_inductives',
+        'print_interps',
+        'print_locale',
+        'print_locales',
+        'print_methods',
+        'print_options',
+        'print_orders',
+        'print_quot_maps',
+        'print_quotconsts',
+        'print_quotients',
+        'print_quotientsQ3',
+        'print_quotmapsQ3',
+        'print_record',
+        'print_rules',
+        'print_simpset',
+        'print_state',
+        'print_statement',
+        'print_syntax',
+        'print_term_bindings',
+        'print_theorems',
+        'print_theory',
+        'print_trans_rules',
+        'smt_status',
+        'thm_deps',
+        'thm_oracles',
+        'thy_deps',
+        'unused_thms',
+        'value',
+        'values',
+        'welcome',
+        'term',
+        'prop',
+        'thm',
+        'typ',
+    ])
+)
+
+find_proof_words = (
+    # title
+    "Proof/counterexample finder",
+    # message
+    "This command is usually used interactively only to find a proof or counter example, "
+    "but only its result should be checked in.",
+    # word(ish) regexps that should be flagged as warnings
+    regexp_of([
+        'nitpick',
+        'nunchaku',
+        'quickcheck',
+        'sledgehammer',
+        'solve_direct',
+        'try',
+        'try0',
+    ])
+)
+
+sorry_words = (
+    # title
+    "Unfinished proof",
+    # message
+    "This command indicates an unfinished or aborted proof.",
+    # word(ish) regexps that should be flagged as warnings
+    regexp_of([
+        '\\\<proof>',
+        'sorry',
```

Comment:
I can see `sorry` being flagged becoming annoying in long running branches such as rt. I agree that it makes sense to have it as a warning by default, but it would be very nice if there was a structured way to configure it per branch.

Reply:
Yes, I should probably add a few options to remove specific classes of matches so that we can configure it at the use site.

- [ ] add options